front: add a modal to manage waypoints in manchette

SharglutDev opened the pull request:
close #8627 
close #8628


> [!WARNING]
> The shift + click doesn't work with firefox (https://github.com/OpenRailAssociation/osrd-ui/issues/510)

Example of testing : 
- Add a scenario 1 with a train with path 1
- Add a scenario 2 with a train with path 1 and others trains with different paths
- Check that by default only the BV/00/'' and map click waypoints are displayed
- Try to unselect all waypoints and see the warning message
- Try to submit while there is the warning and see the animation (the modal shouldn't close)
- Click on `cancel` and verify that the changes you did have not been saved and that manchette is unaltered
- Try the shift + click selection (only on chrome for now) to select and unselect multiple waypoints
- In scenario 2, path 1, hide waypoints in the middle of the path and see the changes in both manchette modes
- Hide a bunch of waypoints starting from the origin included and see the changes in both manchette modes
- When submitting the form, check that an item has been added to local storage
- Go back to scenario 1 and verify that the manchette (from path 1) has not the changes from the same path of scenario 2
- Try deleting the scenario having a saved manchette and verify that the item in local storage is removed
- Repeat the operation with a study having a scenario having a saved manchette and same with a project

SharglutDev committed:
front: add a modal to manage waypoints in manchette

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

Akctarus commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,171 @@
+            <Checkbox
+              small
+              checked={selectedWaypoints.has(index)}
+              onChange={() => {}}
```
```suggestion

```
If onChange is not needed, can we consider removing it ?

SharglutDev replied:
There is a warning in the console if I don't put the onChange in the Checkbox. The logic in the `onClick` can't really be used in the `onChange` because this event doesn't seem to capture keyboard event for the click+shift feature.

SharglutDev replied:
I can leave a comment for that if you want

Akctarus replied:
Yes, it can be useful

SharglutDev replied:
Done in the other commit

Akctarus commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,171 @@
+          />
+        </div>
+        {waypoints.map((waypoint, index) => (
+          <div className="waypoint-item" key={`${waypoint.id}-${index}`}>
```
Isn't `waypoint.id` already unique ? we may not need to concatenate the index.

SharglutDev replied:
I don't think we can have the case for now, but a train could pass to the same op twice if he does a turn around

Akctarus replied:
good point, let's keep it that way then

Akctarus reviewed (commented):
Tested, works well, left two comments

clarani commented on front/src/applications/operationalStudies/views/Project.tsx:
```diff
@@ -61,15 +62,46 @@ export default function Project() {
     deleteItems,
-  } = useMultiSelection<StudyWithScenarios>((studyId) => {
+  } = useMultiSelection<StudyWithScenarios>(async (studyId) => {
     deleteStudy({ projectId: projectId!, studyId });
```
~I think it would be better to delete the study at the end, once the local storage is cleaned~
Sorry, it's better like this I think ! You can resolve this comment

SharglutDev committed:
front: bump osrd-ui v0.0.49

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

clarani commented on front/public/locales/fr/simulation.json:
```diff
@@ -79,5 +79,10 @@
+  "waiting": "Chargement en cours…",
+  "waypointsModal": {
+    "name": "nom",
+    "ch": "CH",
```
Suggestion, maybe we should start to use everywhere secondaryCode instead of CH
```suggestion
    "secondaryCode": "CH",
```

SharglutDev replied:
Good idea, done.

clarani commented on front/src/applications/operationalStudies/Home.tsx:
```diff
@@ -29,24 +30,47 @@ type SortOptions =
+  } = useMultiSelection<ProjectWithStudies | SearchResultItemProject>(async (projectId) => {
+    // For each scenario in the selected projects, clean the local storage if a manchette is saved
+    const { data: studies } = await getStudies({ projectId });
+    if (studies) {
+      const promisedScenarios = studies.results.map(async (study) => {
+        const { data } = await getScenarios({
+          projectId,
+          studyId: study.id,
+        });
+        return data?.results;
+      });
+
+      const scenarios = await Promise.all(promisedScenarios);
+
+      scenarios.flat().forEach((scenario) => {
+        if (scenario) cleanScenarioLocalStorage(scenario.timetable_id);
+      });
+    }
```
Do you think you can move this code in a helper ? In modules/projects/helpers ?
You can use it directly in AddOrEditProject

SharglutDev replied:
Since it's just a temporary code, I was a little lazy to do that but you're right it's better in the meantime :)

Akctarus reviewed (approved):
Tested, LGTM

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,197 @@
+  projectedTrainPath: TrainScheduleBase['path'];
+};
+
+const WaypointsModal = ({
```
Do you think WaypointsPanel would be a better name ?
```suggestion
const WaypointsModal = ({
```

SharglutDev replied:
Fine by me :)

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,197 @@
+import { mmToKm } from 'utils/physics';
+
+type WaypointsModalProps = {
+  waypointsModalOpen: boolean;
```
```suggestion
  waypointsPanelIsOpen: boolean;
```

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,197 @@
+type WaypointsModalProps = {
+  waypointsModalOpen: boolean;
+  setWaypointsModalOpen: (open: boolean) => void;
+  waypoints: NonNullable<PathProperties['operational_points']>;
```
Maybe we can define a new type:
`type OperationalPoint = NonNullable<PathProperties['operational_points']>`

SharglutDev replied:
Thought about it but wasn't sure let's do it :)

SharglutDev replied:
> [!NOTE]
> There is already a type `OperationPoint` in `generatedEditoastApi`, strangely some properties differ from the one in `PathProperties` like one has `part` and the other has `parts`. Maybe there could be a small editoast refacto to harmonize this so we could extend the editoast type with `position`.

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,197 @@
+  const [selectedWaypoints, setSelectedWaypoints] = useState<Set<number>>(
+    new Set(filteredWaypoints.map((_, i) => i))
+  );
+  const [lastSelectedIndex, setLastSelectedIndex] = useState<number | null>(null);
```
Why null and not undefined ? 
```suggestion
  const [lastSelectedIndex, setLastSelectedIndex] = useState<number>();
```

SharglutDev replied:
No specific reason, done.

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -69,7 +69,11 @@ const SimulationResults = ({
+    operationalPoints: projectedOperationalPoints,
+    filteredOperationalPoints,
+    setFilteredOperationalPoints,
+  } = useGetProjectedTrainOperationalPoints(
```
Do you think you can move this hook in the ManchetteWithSpaceTimeChart ?

SharglutDev replied:
Done

SharglutDev replied:
Turns out, `ManchetteWithSpaceTimeChart` is also used in stdcm debug and it doesn't use the same datas so I should revert these changes.

SharglutDev replied:
Reverted the changes :)

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsModal.tsx:
```diff
@@ -0,0 +1,197 @@
+  waypoints: NonNullable<PathProperties['operational_points']>;
+  filteredWaypoints: NonNullable<PathProperties['operational_points']>;
+  setFilteredWaypoints: (waypoints: NonNullable<PathProperties['operational_points']>) => void;
+  projectedTrainPath: TrainScheduleBase['path'];
```
Only a suggestion
```suggestion
  projectionPath: TrainScheduleBase['path'];
```

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/ManchetteMenuButton.tsx:
```diff
@@ -8,7 +8,11 @@ import type { OSRDMenuItem } from 'common/OSRDMenu';
 
-const ManchetteMenuButton = () => {
+type ManchetteMenuButtonProps = {
+  setWaypointsModalOpen: (waypointsModalOpen: boolean) => void;
```
```suggestion
  setWaypointsPanelIsOpen: (waypointsPanelIsOpen: boolean) => void;
```

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/useGetProjectedTrainOperationalPoints.ts:
```diff
@@ -9,16 +11,24 @@ import {
 } from 'common/api/osrdEditoastApi';
+import { useOsrdConfSelectors } from 'common/osrdContext';
+
+const isStation = (chCode: string): boolean => chCode === 'BV' || chCode === '00' || chCode === '';
```
Thanks for this utils function :heart_hands: 
Maybe you can move it in modules/pathfinding ?

SharglutDev replied:
Done

clarani commented on front/src/styles/scss/applications/_operationalStudies.scss:
```diff
@@ -1,8 +1,9 @@
 @import 'operationalStudies/importtrainschedule';
 @import 'operationalStudies/simulationresults';
-@import 'operationalStudies/nge';
+@import 'operationalStudies/waypointsmodal';
```
Suggestion ?
```suggestion
@import 'operationalStudies/waypointsPanel';
```

SharglutDev replied:
Done (also renamed all classes with `panel`)

clarani commented on front/src/modules/simulationResult/types.ts:
```diff
@@ -33,3 +34,11 @@ export type ProjectionData = {
   allTrainsProjected: boolean;
 };
+
+export type WaypointsModalData = {
```
Maybe you can remove this type

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -59,7 +71,9 @@ const ManchetteWithSpaceTimeChartWrapper = ({
             className="inset-0 absolute h-full"
-            spaceOrigin={0}
+            spaceOrigin={
+              (waypointsModalData?.filteredOperationalPoints ?? operationalPoints)[0]?.position || 0
```
```suggestion
              (waypointsModalData?.filteredOperationalPoints ?? operationalPoints).at(0)?.position || 0
```

SharglutDev replied:
Done

SharglutDev committed:
front: fix space time chart scales error

For a brief moment, the computed scales passed to the space time chart can be an empty array.
Add a condition to be sure we render the chart only if there at least of scale.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

clarani commented on front/src/modules/project/helpers/cleanLocalStorageByProject.ts:
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable import/prefer-default-export */
```
I think you can default export this function
```suggestion
```

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -2,33 +2,36 @@ import { useRef, useState } from 'react';
-    part: OperationalPointPart;
-    position: number;
-  }[];
+  operationalPoints: NonNullable<PathProperties['operational_points']>;
```
```suggestion
  operationalPoints: OperationalPoint[];
```

SharglutDev replied:
Done

clarani commented on front/src/modules/simulationResult/components/ManchetteWithSpaceTimeChart/ManchetteWithSpaceTimeChart.tsx:
```diff
@@ -57,16 +67,24 @@ const ManchetteWithSpaceTimeChartWrapper = ({
+            <SpaceTimeChart
+              className="inset-0 absolute h-full"
+              spaceOrigin={
+                (waypointsPanelData?.filteredWaypoints ?? operationalPoints)[0]?.position || 0
```
```suggestion
                (waypointsPanelData?.filteredWaypoints ?? operationalPoints).at(0)?.position || 0
```

SharglutDev replied:
Sorry missed that part in a rebase. Done.

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsPanel.tsx:
```diff
@@ -0,0 +1,194 @@
+    if (event.shiftKey && lastSelectedIndex !== undefined) {
+      const minIndex = Math.min(index, lastSelectedIndex);
+      const maxIndex = Math.max(index, lastSelectedIndex);
+      const newSet = new Set(selectedWaypoints);
```
Move newSet l97 outside the if

SharglutDev replied:
Good point, done

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChart/WaypointsPanel.tsx:
```diff
@@ -0,0 +1,194 @@
+      setSelectedWaypoints(newSet);
+    } else {
+      setSelectedWaypoints((prev) => {
+        const newSet = new Set(prev);
+        if (newSet.has(index)) {
+          newSet.delete(index);
+        } else {
+          newSet.add(index);
+        }
+        return newSet;
```
You don't need the prev here I think :)
```suggestion
        const newSet = new Set(selectedWaypoints); // to move l97
        if (newSet.has(index)) {
          newSet.delete(index);
        } else {
          newSet.add(index);
        }
        setSelectedWaypoints(newSet) // this can even be moved outside the if ... else ...
```

SharglutDev replied:
Done. Because of eslint, had to change a bit the last else.

clarani commented on front/src/styles/scss/applications/operationalStudies/_waypointsPanel.scss:
```diff
@@ -0,0 +1,169 @@
+    color: var(--grey50);
+    background-color: var(--ambiantB15);
+
+    &::after {
+      content: '';
+      position: absolute;
+      bottom: 0.25rem;
+      left: 0;
+      width: 100%;
+      height: 0.0625rem;
+      background-color: var(--grey30);
+    }
```
Why do you implement this like this and not with a bottom border ? :thinking:

SharglutDev replied:
Because on the mockup, this grey line is not at the bottom of the header but 4px above :) (that's why I put a `bottom: 0.25rem`)

![Capture d’écran 2024-10-17 à 15 03 38](https://github.com/user-attachments/assets/517c1a9a-7c3d-4476-a11a-7eaf0ae402fa)

clarani reviewed (commented):
I can still find some "waypointsModal" in the code :) in translations and in ManchetteMenuButton :)

clarani reviewed (approved):
LGTM :white_check_mark:
